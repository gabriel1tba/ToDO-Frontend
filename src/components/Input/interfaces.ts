import { IconBaseProps } from 'react-icons';

export interface IInput extends React.InputHTMLAttributes<HTMLInputElement> {
  icon?: React.ComponentType<IconBaseProps>;
  error: string | undefined;
}
