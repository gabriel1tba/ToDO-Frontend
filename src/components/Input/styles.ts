import styled, { css } from 'styled-components';

import Tooltip from '../Tooltip';

interface WrapperProps {
  isFocused: boolean;
  isFilled: boolean;
  hasError: boolean;
}

export const Wrapper = styled.div<WrapperProps>`
  display: flex;
  align-items: center;

  background: #ffffff;
  border-radius: 10px;
  border: 1px solid #666360;
  padding: 16px;
  width: 100%;
  color: #666360;

  & + div {
    margin-top: 8px;
  }

  ${({ hasError }) =>
    hasError &&
    css`
      border-color: #c53030;
    `}

  ${({ isFocused }) =>
    isFocused &&
    css`
      color: #ff9000;
      border-color: #ff9000;
    `}

  ${({ isFilled }) =>
    isFilled &&
    css`
      color: #ff9000;
      border-color: #ff9000;
    `}

  input {
    flex: 1;
    border: 0;
    background: transparent;

    color: #202124;
    box-shadow: 0 0 0 30px #ffffff inset;

    &:-webkit-autofill {
      -webkit-text-fill-color: #ff9000;
    }

    &::placeholder {
      color: #202124;
    }
  }

  svg {
    margin-right: 16px;
  }
`;

export const Error = styled(Tooltip)`
  height: 20px;
  margin-left: 16px;

  > svg {
    margin: 0;
  }

  span {
    background: #c53030;
    color: #fff;
    font-size: 16px;

    &::before {
      border-color: #c53030 transparent;
    }
  }
`;
