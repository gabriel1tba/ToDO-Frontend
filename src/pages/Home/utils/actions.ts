import { TTodo } from 'services/TodoService/interfaces';

export enum Types {
  GET_TODOS = 'GET_TODOS',
  CREATE_TODO = 'CREATE_TODO',
  UPDATE_TODO = 'UPDATE_TODO',
  DELETE_TODO = 'DELETE_TODO',
}

interface GetTodos {
  type: Types.GET_TODOS;
  payload: TTodo[];
}

interface CreateTodo {
  type: Types.CREATE_TODO;
  payload: TTodo;
}

interface UpdateTodo {
  type: Types.UPDATE_TODO;
  payload: TTodo;
}

interface DeleteTodo {
  type: Types.DELETE_TODO;
  payload: string;
}

export type Action = GetTodos | CreateTodo | UpdateTodo | DeleteTodo;

export const getTodosAction = (todos: TTodo[]): GetTodos => ({
  type: Types.GET_TODOS,
  payload: todos,
});

export const createTodoAction = (todo: TTodo): CreateTodo => ({
  type: Types.CREATE_TODO,
  payload: todo,
});

export const updateTodoAction = (todo: TTodo): UpdateTodo => ({
  type: Types.UPDATE_TODO,
  payload: todo,
});

export const deleteTodoAction = (id: string): DeleteTodo => ({
  type: Types.DELETE_TODO,
  payload: id,
});
