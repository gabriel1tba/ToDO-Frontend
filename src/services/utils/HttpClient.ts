import axios from 'axios';

const HttpClient = axios.create({
  baseURL: `${process.env.REACT_APP_API_HOST}:${process.env.REACT_APP_API_PORT}`,
});

export default HttpClient;
