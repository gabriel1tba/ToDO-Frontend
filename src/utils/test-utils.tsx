import { render, RenderOptions } from '@testing-library/react';
import { TodoProvider } from 'context/todos';

interface IWrapperProvider {
  children: React.ReactNode;
}

const WrapperProvider = ({ children }: IWrapperProvider) => (
  <TodoProvider>{children}</TodoProvider>
);

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(<WrapperProvider>{ui}</WrapperProvider>, { ...options });

export * from '@testing-library/react';
export { customRender as render };
