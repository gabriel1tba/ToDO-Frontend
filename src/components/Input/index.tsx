import { useRef, useState, forwardRef, ForwardRefRenderFunction } from 'react';
import { FiAlertCircle } from 'react-icons/fi';
import { IconType } from 'react-icons';

import * as S from './styles';

export interface IInput extends React.InputHTMLAttributes<HTMLInputElement> {
  icon?: IconType;
  error: string | undefined;
}

const Input: ForwardRefRenderFunction<HTMLInputElement, IInput> = (
  { icon: Icon, error, ...props },
  ref
) => {
  const [isFocused, setIsFocued] = useState(false);
  const [isFilled, setIsFilled] = useState(false);

  const inputRef = useRef({} as HTMLInputElement);

  const handleInputFocus = () => {
    setIsFocued(true);
  };

  const handleInputBlur = () => {
    setIsFocued(false);

    inputRef.current.value ? setIsFilled(true) : setIsFilled(false);
  };

  return (
    <S.Wrapper
      hasError={!!error}
      isFocused={isFocused}
      isFilled={isFilled}
      onFocus={handleInputFocus}
      onBlur={handleInputBlur}
      ref={inputRef}
    >
      {Icon && <Icon size={20} />}

      <input ref={ref} {...props} />
      {error && (
        <S.Error title={error}>
          <FiAlertCircle />
        </S.Error>
      )}
    </S.Wrapper>
  );
};

export default forwardRef(Input);
