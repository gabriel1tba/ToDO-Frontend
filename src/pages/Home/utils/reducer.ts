import { TTodo } from 'services/TodoService/interfaces';

import { Types, Action } from './actions';

const todosReducer = (state: TTodo[], action: Action): TTodo[] => {
  switch (action.type) {
    case Types.GET_TODOS: {
      return action.payload;
    }

    case Types.CREATE_TODO: {
      return [...state, action.payload];
    }

    case Types.UPDATE_TODO: {
      return state.map((todo) => {
        if (action.payload.id === todo.id) {
          todo = { ...action.payload, created_at: todo.created_at };
        }
        return todo;
      });
    }

    case Types.DELETE_TODO: {
      return state.filter((todo) => {
        return todo.id !== action.payload;
      });
    }

    default:
      return state;
  }
};

export default todosReducer;
