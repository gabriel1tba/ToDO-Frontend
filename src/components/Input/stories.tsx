import { Story, Meta } from '@storybook/react';
import { FaUser } from 'react-icons/fa';

import Input, { IInput } from '.';

export default {
  title: 'Form/Input',
  component: Input,
  args: {
    label: 'E-mail',
    name: 'email',
    initialValue: '',
    placeholder: 'john.cage@gmail.com',
  },
  argTypes: {
    name: {
      type: 'string',
    },
    icon: {
      type: 'function',
    },
    error: {
      type: 'string',
    },
    onChange: {
      action: 'change',
    },
  },
} as Meta;

export const Default: Story<IInput> = (args) => <Input {...args} />;

export const withIcon: Story<IInput> = (args) => <Input {...args} />;

withIcon.args = {
  icon: FaUser,
};

export const withError: Story<IInput> = (args) => <Input {...args} />;

withError.args = {
  error: 'Ops...something is wrong',
};
