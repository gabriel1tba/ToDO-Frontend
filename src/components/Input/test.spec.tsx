import userEvent from '@testing-library/user-event';
import { render, screen } from 'utils/test-utils';

import Input from '.';

describe('<Input />', () => {
  it('should render the Input correctly', () => {
    const { container } = render(
      <Input name="email" error="" placeholder="Digite seu e-mail" />,
    );

    expect(
      screen.getByPlaceholderText(/digite seu e-mail/i),
    ).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should write in the input', async () => {
    render(<Input name="email" error="" placeholder="Digite seu e-mail" />);

    userEvent.type(
      screen.getByPlaceholderText(/digite seu e-mail/i),
      'Testes finalizados',
    );

    expect(screen.getByPlaceholderText(/digite seu e-mail/i)).toHaveValue(
      'Testes finalizados',
    );
  });

  it('should render erros', () => {
    render(<Input name="email" error="Digite o e-mail corretamente" />);

    expect(
      screen.getByText(/digite o e-mail corretamente/i),
    ).toBeInTheDocument();
  });
});
