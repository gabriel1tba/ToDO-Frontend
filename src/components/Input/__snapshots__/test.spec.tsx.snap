// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should render the Input correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #FFFFFF;
  border-radius: 10px;
  border: 1px solid #424242;
  padding: 14px;
  width: 100%;
  color: #424242;
}

.c0 + div {
  margin-top: 10px;
}

.c0 input {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  border: 0;
  background: transparent;
  color: #202124;
  box-shadow: 0 0 0 30px #FFFFFF inset;
}

.c0 input:-webkit-autofill {
  -webkit-text-fill-color: #ff9000;
}

.c0 input::-webkit-input-placeholder {
  color: #202124;
}

.c0 input::-moz-placeholder {
  color: #202124;
}

.c0 input:-ms-input-placeholder {
  color: #202124;
}

.c0 input::placeholder {
  color: #202124;
}

.c0 svg {
  margin-right: 16px;
}

<div
  class="c0"
>
  <input
    name="email"
    placeholder="Digite seu e-mail"
  />
  
</div>
`;
